From 31620be4fa066fdd5a742b3d3acc30973ad54e3b Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Wed, 7 Nov 2018 14:14:11 +0800
Subject: [PATCH 5039/5242] MLK-20259 ARM: imx: add soc revision check for
 i.MX7ULP

commit  31451d7fa2fbfd78c9a800fa51a9924ac22747ab from
https://source.codeaurora.org/external/imx/linux-imx.git

i.MX7ULP SoC revision is available from B0, the SIM_JTAG_ID
register bit[31:28] indicates SoC revision as below:

4b'0001        B0
4b'0010        B1

This register is NOT available on A0, tested on B1 chip
as below:

root@imx7ulpevk:~# cat /sys/devices/soc0/revision
2.1

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Tested-by: Ye Li <ye.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/mach-imx7ulp.c |   35 +++++++++++++++++++++++++++++++++++
 1 file changed, 35 insertions(+)

diff --git a/arch/arm/mach-imx/mach-imx7ulp.c b/arch/arm/mach-imx/mach-imx7ulp.c
index 2d916e1..bf385b4 100644
--- a/arch/arm/mach-imx/mach-imx7ulp.c
+++ b/arch/arm/mach-imx/mach-imx7ulp.c
@@ -8,7 +8,9 @@
  */
 
 #include <linux/irqchip.h>
+#include <linux/mfd/syscon.h>
 #include <linux/of_platform.h>
+#include <linux/regmap.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/map.h>
 
@@ -25,9 +27,42 @@
 	mx7ulp_aips_map_entry(5, MT_DEVICE),
 };
 
+#define SIM_JTAG_ID_REG		0x8c
+
 static void __init imx7ulp_init_machine(void)
 {
 	struct device *parent;
+	struct regmap *sim;
+	u32 revision;
+	int ret;
+
+	sim = syscon_regmap_lookup_by_compatible("fsl,imx7ulp-sim");
+	if (IS_ERR(sim)) {
+		pr_err("failed to find fsl,imx7ulp-sim regmap!\n");
+		return;
+	}
+
+	ret = regmap_read(sim, SIM_JTAG_ID_REG, &revision);
+	if (ret)
+		pr_err("failed to read sim regmap!\n");
+
+	/*
+	 * bit[31:28] of JTAG_ID register defines revision
+	 * as below from B0:
+	 * 0001        B0
+	 * 0010        B1
+	 */
+	switch (revision >> 28) {
+	case 1:
+		imx_set_soc_revision(IMX_CHIP_REVISION_2_0);
+		break;
+	case 2:
+		imx_set_soc_revision(IMX_CHIP_REVISION_2_1);
+		break;
+	default:
+		imx_set_soc_revision(IMX_CHIP_REVISION_1_0);
+		break;
+	}
 
 	parent = imx_soc_device_init();
 	if (parent == NULL)
-- 
1.7.9.5

